File: 백준/Gold/15683. 감시/감시.py
import copy

n, m = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]
#6은 벽, 0은 빈칸, 1-5는 CCTV의 번호
answer = int(1e9) # 사각 지대의 최소 크기
cctv = []
mode = [
    [],  # cctv가 회전할 때마다 가능한 방향을 0을 동쪽으로 정의 후에 리스트로 저장
    [[0], [1], [2], [3]], #1번 cctv
    [[0, 2], [1, 3]], #2번 cctv
    [[0, 1], [1, 2], [2, 3], [0, 3]], #3번 cctv
    [[0, 1, 2], [0, 1, 3], [1, 2, 3], [0, 2, 3]], #4번 cctv
    [[0, 1, 2, 3]], #5번 cctv
]

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]

for i in range(n):
  for j in range(m):
    if 1 <= board[i][j] <= 5:
      cctv.append((board[i][j], i, j))

def fill(board, mode, x, y):
    for i in mode:
        nx = x
        ny = y
        while True:
            nx += dx[i]
            ny += dy[i]
            if not 0 <= nx < n or not 0 <= ny < m: # 범위를 벗어나면 break
                break
            if board[nx][ny] == 6: # 벽을 마주치면 멈춤
                break
            elif board[nx][ny] == 0: #빈공간이면 7을 할당하여 감시 가능한 영역으로 마킹
                board[nx][ny] = 7

def dfs(depth, arr):
    global answer

    if depth == len(cctv) :
        count = 0
        for i in range(n) :
            count += arr[i].count(0)
        answer = min(answer, count)
        return

    temp = copy.deepcopy(arr)
    cctv_num, x, y = cctv[depth]
    for i in mode[cctv_num]:
        fill(temp, i, x, y)
        dfs(depth+1, temp)
        temp = copy.deepcopy(arr)


dfs(0, board)
print(answer)

File: 백준/Gold/15683. 감시/감시.swift
import Foundation

let readline = readLine()!.split(separator:" ").map { Int($0)! }
let (n, m) = (readline[0], readline[1])
var board:[[Int]] = []
var cctvs:[(Int, Int)] = []  // cctv의 좌표를 저장하는 배열
let dx = [1, 0, -1, 0]
let dy = [0, 1, 0, -1]
// 우 하 좌 상 순서 offset 배열
var answer = Int.max

for i in 0..<n {
  board.append(readLine()!.split(separator:" ").map { Int($0)! })
  for j in 0..<m {
    if Array(1...5).contains(board[i][j]) {
      cctvs.append((i, j))
    }
  }
}

func isValidCoord(_ x: Int, _ y: Int) -> Bool {    // 유효한 좌표인지를 판단하는 함수
  return 0 <= x && x < n && 0 <= y && y < m
}

func setMonitor(_ cctv_x: Int, _ cctv_y: Int, _ dir: Int) {
  var x = cctv_x
  var y = cctv_y
  let dir = dir % 4   // dir가 인덱스 범위를 벗어나면 다시 인덱스 범위에 속하도록 만들어주기 위해 계산하는 식

  while true {  // 벽을 마주칠때까지 반복
    x += dx[dir]
    y += dy[dir]

    if !isValidCoord(x, y) || board[x][y] == 6 { break }
    if board2[x][y] == 0 {
      board2[x][y] = -1
    }
  }
}

var board2 = board   // 매번 케이스에 따라서 갯수가 달라지므로 복사에 사용할 배열
for i in 0..<Int(pow(4.0, Double(cctvs.count))) {     // 4^cctv의 갯수가 모든 경우의 수
  var brute = i
  board2 = board

  for j in 0..<cctvs.count {
    let dir = brute % 4
    brute /= 4
    let (x,y) = cctvs[j]
    let type = board[x][y]

        switch type {
        case 1:
            setMonitor(x, y, dir)
        case 2:
            setMonitor(x, y, dir)
            setMonitor(x, y, dir + 2)
        case 3:
            setMonitor(x, y, dir)
            setMonitor(x, y, dir + 1)
        case 4:
            setMonitor(x, y, dir)
            setMonitor(x, y, dir + 1)
            setMonitor(x, y, dir + 2)
        case 5:
            setMonitor(x, y, dir)
            setMonitor(x, y, dir + 1)
            setMonitor(x, y, dir + 2)
            setMonitor(x, y, dir + 3)
        default:
            break
        }
    }

    var blindCount = 0
    for i in 0..<n {
        for j in 0..<m {
            if board2[i][j] == 0 {
                blindCount += 1
            }
        }
    }

    answer = min(answer, blindCount)
}

print(answer)